File: services/api/internal/model/message.go
package model

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

// Message represents a message in the system.
type Message struct {
	ID             uuid.UUID        `json:"-"`
	SenderID       uuid.UUID        `json:"-"`
	SenderDeviceID *uuid.UUID       `json:"-"`
	GroupID        *uuid.UUID       `json:"-"`
	RecipientID    *uuid.UUID       `json:"-"`
	Content        *string          `json:"-"`
	MessageType    string           `json:"-"`
	Lat            *float64         `json:"-"` // extracted from PostGIS location
	Lng            *float64         `json:"-"`
	Metadata       *json.RawMessage `json:"-"`
	CreatedAt      time.Time        `json:"-"`
}

// Attachment represents a file attached to a message.
type Attachment struct {
	ID          uuid.UUID `json:"-"`
	MessageID   uuid.UUID `json:"-"`
	Filename    string    `json:"-"`
	ContentType string    `json:"-"`
	SizeBytes   int64     `json:"-"`
	ObjectKey   string    `json:"-"`
	CreatedAt   time.Time `json:"-"`
}

// MessageResponse is the JSON representation returned by the API.
type MessageResponse struct {
	ID          uuid.UUID            `json:"id"`
	SenderID    uuid.UUID            `json:"sender_id"`
	Username    string               `json:"username"`
	DisplayName string               `json:"display_name"`
	GroupID     *uuid.UUID           `json:"group_id,omitempty"`
	RecipientID *uuid.UUID           `json:"recipient_id,omitempty"`
	Content     string               `json:"content"`
	MessageType string               `json:"message_type"`
	Lat         *float64             `json:"lat,omitempty"`
	Lng         *float64             `json:"lng,omitempty"`
	Metadata    *json.RawMessage     `json:"metadata,omitempty"`
	Attachments []AttachmentResponse `json:"attachments"`
	CreatedAt   time.Time            `json:"created_at"`
}

// AttachmentResponse is the JSON representation of an attachment.
type AttachmentResponse struct {
	ID          uuid.UUID `json:"id"`
	Filename    string    `json:"filename"`
	ContentType string    `json:"content_type"`
	SizeBytes   int64     `json:"size_bytes"`
	CreatedAt   time.Time `json:"created_at"`
}

// ToResponse converts an Attachment to its API response.
func (a *Attachment) ToResponse() AttachmentResponse {
	return AttachmentResponse{
		ID:          a.ID,
		Filename:    a.Filename,
		ContentType: a.ContentType,
		SizeBytes:   a.SizeBytes,
		CreatedAt:   a.CreatedAt,
	}
}

// MessageWithUser is a join result containing message + sender details.
type MessageWithUser struct {
	Message
	Username    string
	DisplayName *string
	Attachments []Attachment
}

// ToResponse converts a MessageWithUser to its API response.
func (m *MessageWithUser) ToResponse() MessageResponse {
	content := ""
	if m.Content != nil {
		content = *m.Content
	}
	displayName := ""
	if m.DisplayName != nil {
		displayName = *m.DisplayName
	}

	atts := make([]AttachmentResponse, len(m.Attachments))
	for i, a := range m.Attachments {
		atts[i] = a.ToResponse()
	}

	return MessageResponse{
		ID:          m.ID,
		SenderID:    m.SenderID,
		Username:    m.Username,
		DisplayName: displayName,
		GroupID:     m.GroupID,
		RecipientID: m.RecipientID,
		Content:     content,
		MessageType: m.MessageType,
		Lat:         m.Lat,
		Lng:         m.Lng,
		Metadata:    m.Metadata,
		Attachments: atts,
		CreatedAt:   m.CreatedAt,
	}
}

File: clients/ios/SitAware/Models/Message.swift
import Foundation

// MARK: - Attachment

struct Attachment: Codable, Sendable, Identifiable {
    let id: String
    let filename: String
    let contentType: String
    let sizeBytes: Int
    let createdAt: String

    /// Whether this attachment is an image type.
    var isImage: Bool {
        contentType.hasPrefix("image/")
    }

    /// Human-readable file size.
    var formattedSize: String {
        let formatter = ByteCountFormatter()
        formatter.countStyle = .file
        return formatter.string(fromByteCount: Int64(sizeBytes))
    }
}

// MARK: - EXIF Location

/// GPS location extracted from an image attachment's EXIF data.
struct ExifLocation: Codable, Sendable {
    let attachmentId: String
    let lat: Double
    let lng: Double
    var altitude: Double?
    var takenAt: String?
}

// MARK: - Message Metadata

/// Typed message metadata — shape depends on message_type.
struct MessageMetadata: Codable, Sendable {
    var exifLocations: [ExifLocation]?
    var drawingId: String?

    // Catch-all for unknown keys (GPX data stored as GeoJSON, etc.)
    private var additionalProperties: [String: AnyCodable]?

    enum CodingKeys: String, CodingKey {
        case exifLocations
        case drawingId
    }
}

// MARK: - Message

struct MessageResponse: Codable, Sendable, Identifiable {
    let id: String
    let senderId: String
    let username: String
    let displayName: String
    var groupId: String?
    var recipientId: String?
    let content: String
    let messageType: String
    var lat: Double?
    var lng: Double?
    var metadata: MessageMetadata?
    let attachments: [Attachment]
    let createdAt: String
}

// MARK: - WebSocket Message

/// Server -> Client: a new message via WebSocket (same shape as MessageResponse).
typealias WSMessageNew = MessageResponse

// MARK: - Conversations

/// A user the caller has DM history with.
struct DMConversationPartner: Codable, Sendable {
    let userId: String
    let username: String
    let displayName: String
}

/// Conversation list item — either a group or a DM partner.
struct Conversation: Sendable, Identifiable, Equatable {
    let id: String
    let type: ConversationType
    let name: String
    var lastMessage: MessageResponse?

    static func == (lhs: Conversation, rhs: Conversation) -> Bool {
        lhs.id == rhs.id && lhs.type == rhs.type
    }
}

enum ConversationType: String, Sendable {
    case group
    case direct
}

// MARK: - AnyCodable Helper

/// Minimal type-erased Codable wrapper for arbitrary JSON values.
struct AnyCodable: Codable, Sendable {
    let value: Any

    init(_ value: Any) {
        self.value = value
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if let string = try? container.decode(String.self) {
            value = string
        } else if let int = try? container.decode(Int.self) {
            value = int
        } else if let double = try? container.decode(Double.self) {
            value = double
        } else if let bool = try? container.decode(Bool.self) {
            value = bool
        } else if let array = try? container.decode([AnyCodable].self) {
            value = array.map(\.value)
        } else if let dict = try? container.decode([String: AnyCodable].self) {
            value = dict.mapValues(\.value)
        } else {
            value = NSNull()
        }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch value {
        case let string as String: try container.encode(string)
        case let int as Int: try container.encode(int)
        case let double as Double: try container.encode(double)
        case let bool as Bool: try container.encode(bool)
        default: try container.encodeNil()
        }
    }
}
